addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductScreen.js

```diff
@@ -229,36 +229,40 @@ odoo.define('point_of_sale.ProductScreen', function(require) {
-            if (!product) {
-                // find the barcode in the backend
-                let foundProductIds = [];
-                try {
-                    foundProductIds = await this.rpc({
-                        model: 'product.product',
-                        method: 'search',
-                        args: [[['barcode', '=', code.base_code], ['sale_ok', '=', true]]],
-                        context: this.env.session.user_context,
+            const barcode = code.base_code;
+            const product = this.env.pos.db.get_product_by_barcode(barcode) 
+                             || await this._findProduct(barcode) 
+                             || this._barcodeErrorAction(code);
+            return product;
+        }
+        async _findProduct(barcode) {
+            // find the barcode in the backend
+            let foundProductIds = [];
+            let product;
+            try {
+                foundProductIds = await this.rpc({
+                    model: 'pos.session',
+                    method: 'find_product_by_barcode',
+                    args: [odoo.pos_session_id, barcode],
+                    context: this.env.session.user_context,
+                });
+            } catch (error) {
+                if (isConnectionError(error)) {
+                    return this.showPopup('OfflineErrorPopup', {
+                        title: this.env._t('Network Error'),
+                        body: this.env._t("Product is not loaded. Tried loading the product from the server but there is a network error."),
                     });
-                } catch (error) {
-                    if (isConnectionError(error)) {
-                        return this.showPopup('OfflineErrorPopup', {
-                            title: this.env._t('Network Error'),
-                            body: this.env._t("Product is not loaded. Tried loading the product from the server but there is a network error."),
-                        });
-                    } else {
-                        throw error;
-                    }
-                }
-                if (foundProductIds.length) {
-                    await this.env.pos._addProducts(foundProductIds);
-                    // assume that the result is unique.
-                    product = this.env.pos.db.get_product_by_id(foundProductIds[0]);
                 } else {
-                    return this._barcodeErrorAction(code);
+                    throw error;
                 }
             }
-            return product
+            if (foundProductIds) {
+                const { product: foundProduct, packaging: foundPackaging = [] } = foundProductIds;
+                await this.env.pos._loadProductProduct(foundProduct);
+                this.env.pos.db.add_packagings(foundPackaging);
+                product = this.env.pos.db.get_product_by_id(foundProduct[0].id);
+            }
+            return product;
```

Comment:
directly put this logic inside the try-catch. Also, you probably don't need the "foundProductIds" var, just destructure immediately the result.

Reply:
I put it inside the try-catch, but we need the var as it can return None.